### DISCUSSION TITLE
TextCategorizer with transformer (no fine-tuning)?

### DISCUSSION THREAD

**Discussion by apgr0ss** [explosion/spaCy#8120] [Help: Coding & Implementations] [answered]
Hi,

I would like to train a TextCat model with one of the pre-trained HuggingFace transformers. However, I'm still waiting for gpu access, so having to also fine-tune the transformer causes training to be incredibly slow.

To remedy this, I was thinking as a benchmark model just run the text through the pre-trained transformer (no fine-tuning), collect the sentence embeddings and pass them into a logistic regression.

Can anyone provide a set of general steps to follow in order to do this in spaCy? Thanks!

**Comment by polm** [comment, marked as answer]
You can just freeze the transformer layer to avoid fine-tuning it. You can read about freezing layers [here](https://spacy.io/usage/training#:~:text=if%20you%20don%E2%80%99t%20want%20a%20component%20to%20be%20updated%2C%20you%20can%20freeze%20it%20by%20adding%20it%20to%20the%20frozen_components%20list%20in%20the%20%5Btraining%5D%20block.). 

However, I don't think this will solve your problem. Even if you're not fine-tuning it, running a Transformer on CPU will be fantastically slow. Until you get GPU access it might be a good idea to get a CPU pipeline up and running to make sure everything else is OK and establish a baseline. It may even turn out the CPU models are good enough for your use case.

**Reply by apgr0ss** [reply]
Thanks for the response Paul. 

Just for my own knowledge, could you provide a brief explanation for why the transformer would run slowly on cpu even if I'm just using it for prediction?

**Reply by polm** [reply]
Transformers use many self-attention layers, whether you're training or just doing inference. Attention is basically matrix multiplication, which is what GPUs are good at. 

I had a hard time finding numbers on this, but [this Huggingface post](https://medium.com/huggingface/benchmarking-transformers-pytorch-and-tensorflow-e2917fb891c2) suggests inference on CPU is roughly 20 times slower than GPU, which is honestly better than I expected.
